### PR TITLE
Restructuring functionalities to Export an API/API Product with new schemas with relevant DTOs for APICTL

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Fixes: https://github.com/wso2/product-apim-tooling/issues/324
Fixes: https://github.com/wso2/product-apim-tooling/issues/158

## Goals
Restructure the code segments related to exporting an API/API Product by introducing new schemas via writing required DTOs.

## Approach
- Created/Modified Publisher REST API resources to export an API/API Product. The resources are as follows.
  1. GET apis/export
  2. GET api-products/export
- The old implementations of exporting APIs and API Products have been modified to write DTOs to the exported files, instead of models.
- The new structure of an Exported API are as follows (assumed that the files are exported in .yaml format).
   ```
   └── PizzaShackAPI-1.0.0
       ├── api_params.yaml
       ├── Client-certificates
       │   ├── Alias1.crt
       │   ├── Alias2.crt
       │   └── client_certificates.yaml
       ├── Docs
       │   ├── Dco4
       │   │   ├── mydoc.pdf
       │   │   └── document.yaml
       │   ├── Doc1
       │   │   ├── Doc1
       │   │   └── document.yaml
       │   ├── Doc2
       │   │   ├── Doc2
       │   │   └── document.yaml
       │   └── Doc3
       │       └── document.yaml
       ├── Endpoint-certificates
       │   ├── Alias3.crt
       │   ├── Alias4.crt
       │   └── endpoint_certificates.yaml
       ├── Meta-information
       │   ├── api.yaml
       │   └── swagger.yaml
       └── Sequences
           ├── fault-sequence
           │   ├── sequence.yaml
           │   └── debug_json_fault.xml
           ├── in-sequence
           │   ├── sequence.yaml
           │   └── Custom
           │       └── sample-n-sequence.xml
           └── out-sequence
                      ├── sequence.yaml
                      └── log_out_message.xml
   ```
- The structure of an API Product is not changes, except the dependent APIs will have the above structure.
- Each data file (schema) will have 3 main fields named as **data**, **version** and **type** as shown below.
   ```
   type: api
   version: v4
   data:
     id: 616794a8-b5db-4bb9-b768-5dbbed836f17
     name: PizzaShackAPI
     description: This is a simple API for Pizza Shack online pizza delivery store.
     context: /pizzashack
     version: 1.0.0
     provider: admin
     lifeCycleStatus: PUBLISHED
     ---------------------------
     ---------------------------
   ```

## User stories
The users can use the newly introduced Publisher REST API resources to export an API/API Product with the new schema structure.

## Test environment
- JDK 1.8.0_251
- Ubuntu 20.04.1 LTS

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java [hunk 1]
```diff
@@ -0,0 +1,1087 @@
...
+                Collection inFlow = (org.wso2.carbon.registry.api.Collection) registry.get(soapToRestBaseUrl
+                        + RegistryConstants.PATH_SEPARATOR + IN);
+                Collection outFlow = (org.wso2.carbon.registry.api.Collection) registry.get(soapToRestBaseUrl
+                        + RegistryConstants.PATH_SEPARATOR + OUT);
+
+                CommonUtil.createDirectory(archivePath + File.separator + SOAPTOREST + File.separator + IN);
+                CommonUtil.createDirectory(archivePath + File.separator + SOAPTOREST + File.separator + OUT);
+                if (inFlow != null) {
+                    for (String inFlowPath : inFlow.getChildren()) {
+                        inputStream = registry.get(inFlowPath).getContentStream();
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:342]: Any reason for not using `try with resources` pattern here, we can avoid needing to close the stream.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java [hunk 2]
```diff
@@ -0,0 +1,1087 @@
...
+                                + File.separator + IN +
+                                inFlowPath.substring(inFlowPath.lastIndexOf(RegistryConstants.PATH_SEPARATOR)));
+                        IOUtils.copy(inputStream, outputStream);
+                        IOUtils.closeQuietly(inputStream);
+                        IOUtils.closeQuietly(outputStream);
+                    }
+                }
+                if (outFlow != null) {
+                    for (String outFlowPath : outFlow.getChildren()) {
+                        inputStream = registry.get(outFlowPath).getContentStream();
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:353]: Any reason for not using try with resources pattern here, we can avoid needing to close the stream.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java [hunk 3]
```diff
@@ -0,0 +1,1087 @@
...
+            throw new APIImportExportException(errorMsg, e);
+        }
+
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        JsonArray certificatesList = new JsonArray();
+        certificateMetadataDTOS.forEach(metadataDTO -> {
+            ByteArrayInputStream certificate = null;
+            try {
+                certificate = certificateManager.getCertificateContent(metadataDTO.getAlias());
+                certificate.close();
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:763]: `ByteArrayInputStream certificate` is accessed after it is closed. Better to use `try whit resources` to ensure that this is not required.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java [hunk 4]
```diff
@@ -0,0 +1,1087 @@
...
+                    JSONObject endpointJSON = new JSONObject(endpointConfig.getJSONObject(type).toString());
+                    try {
+                        String urlValue = endpointJSON.get(APIConstants.API_DATA_URL).toString();
+                        urls.add(urlValue);
+                    } catch (JSONException ex) {
+                        log.error("Endpoint URL extraction from endpoint JSON object failed in API: " + apiName, ex);
+                    }
+                }
+            } catch (JSONException ex) {
+                log.info("Endpoint type: " + type + " not found in API: " + apiName);
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:827]: Should this scenario be ignored?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java [hunk 5]
```diff
@@ -0,0 +1,1087 @@
...
+     * @param clientCertificateDTOs client certificates list DTOs
+     * @param certDirectoryPath     directory path to export the certificates
+     * @return list of certificate detail JSON objects
+     */
+    private static JsonArray getClientCertificateContentAndMetaData(List<ClientCertificateDTO> clientCertificateDTOs,
+                                                                    String certDirectoryPath) {
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        JsonArray certificatesList = new JsonArray();
+        clientCertificateDTOs.forEach(metadataDTO -> {
+            ByteArrayInputStream certificate = null;
```
Comment (uvindra) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:944]: Use the stream with `try with resources`

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 4 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java:827]: No, this should be thrown as an exception. Will fix it. Thanks!

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 1 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9406/commits/150d9abd9d4eb25c7f1217fc53f6143f94167542

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 2 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9406/commits/150d9abd9d4eb25c7f1217fc53f6143f94167542

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 3 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9406/commits/150d9abd9d4eb25c7f1217fc53f6143f94167542

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 4 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9406/commits/150d9abd9d4eb25c7f1217fc53f6143f94167542

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java (hunk 5 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/ExportUtils.java]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9406/commits/150d9abd9d4eb25c7f1217fc53f6143f94167542

**Review by uvindra — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java [hunk 6]
```diff
@@ -346,4 +349,16 @@ public static String getSOAPOperation() {
                 + "\"required\":false,\"in\":\"header\"}],\"responses\":{\"200\":{\"description\":\"OK\"}}," +
                 "\"security\":[{\"default\":[]}],\"consumes\":[\"text/xml\",\"application/soap+xml\"]}}}";
     }
+
+    /**
+     * This method retrieves the Swagger Definition for an API to be displayed
+     * @param api API
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java:355]: Missing apiProvider in the java docs

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java [hunk 7]
```diff
@@ -346,4 +349,16 @@ public static String getSOAPOperation() {
...
+     * @return String
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java:356]: Please define what is return by the String value as well

Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java:356]: Fixed via https://github.com/wso2/carbon-apimgt/commit/eb9b7181c0f6cb5b9624602d91d28a7479bf551b

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java (hunk 6 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/utils/RestApiPublisherUtils.java:355]: Fixed via https://github.com/wso2/carbon-apimgt/commit/eb9b7181c0f6cb5b9624602d91d28a7479bf551b

**Pull request merged**